File: javascript/linked_list/287. Find the Duplicate Number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findDuplicate = function(nums) {
    let fast = 0, slow = 0, slow2 = 0;
    do {
        slow = nums[slow];
        fast = nums[nums[fast]];
    } while (fast !== slow);

    while (slow !== slow2) {
        slow = nums[slow];
        slow2 = nums[slow2];
    }
    return slow;
};

File: python/linked_list/287. Find the Duplicate Number.py
'''
Floyd's Cycle Detection
Time: O(n)
Space: O(1)

Thank values as pointers
0 =p=> 1 ==> 2 ==> 3 ==> ==> 4 ==> 5 =x=> 1
c: cycle
2 * slow = fast 
2(p + c - x) = p + c - x + c
2p + 2c - 2x = p + 2c - x
p = x
'''

class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        slow = fast = slow2 = 0
        while True:
            slow = nums[slow]  # x
            fast = nums[nums[fast]]  # 2x
            if slow == fast:
                break

        while True:
            slow = nums[slow]
            slow2 = nums[slow2]
            if slow == slow2:
                return slow

